tcpdump.1.in

```diff
@@ -1442,7 +1442,7 @@ Recall the structure of a TCP header without options:
 -----------------------------------------------------------------
 |                     acknowledgment number                     |
 -----------------------------------------------------------------
-|  HL   | rsvd  |C|E|U|A|P|R|S|F|        window size            |
+|  HL   |rsvd |A|C|E|U|A|P|R|S|F|        window size            |
```

Comment:
This packet diagram cannot convey the meaning using "A" for two different flags. Either it will have to use more characters per bit, or the flags field needs be unpacked only in a separate 12-bit diagram.

Reply:
Something like?
![flags](https://github.com/the-tcpdump-group/tcpdump/assets/2846004/3e71f289-52d8-4f08-9f5e-fd422fef08ae)

Reply:
And ` W (CWR)` in the man page should be ` C (CWR)`.

Reply:
> And ` W (CWR)` in the man page should be ` C (CWR)`.

No, bad idea. It's 'W' in the code.

Reply:
Vertical labels could work, yes.

Reply:
So - variant one, use the single-char appreviations here that tcpdump would show in the output:
|A|W|E|U|.|P|R|S|F

or expand to 3 lines, and in both cases have  a short table translating the single char to RFC abbreviations?

Reply:
```
 There are 9 bits in the control bits section of the TCP header:
 .I AE | CWR(W) | ECE | URG | ACK(.) | PSH | RST | SYN | FIN
 
|  HL   |rsvd |A|W|E|U|.|P|R|S|F|        window size            |
```

Reply:
> Vertical labels could work, yes.

Adapted from a wikipedia diagram.

Reply:
Use of `.` to indicate `ACK` is a peculiarity of the protocol analyzer software; the motivation behind that may or may not be lost in time, but in any case the discrepancy has fossilized and it would be wrong to fix it in the software. It would also be wrong to diverge from the normative specification's version of the packet diagrams in the man page. So the best balance seems to be making the man page closer to the spec and preserving the traditional remarks such as "The `.' means the ACK flag was set." and the likes.

Reply:
To avoid confusion 'A' with ACK, could the new flag for 'Accurate ECN' be 'N'?

Reply:
Do you mean in the `[SF.PUE]` output syntax? It is not difficult to imagine how "A" could easily confuse less experienced users. Bit 7 used to mean NS in the past, and N is the only character that appears in both "Accurate ECN" and "NS" (not in "AE" though). Or perhaps it could be just "7", which would also simplify representation of the remaining three bits. Perhaps it would be best to bring this up on the mailing list.

But the man page should be consistent with the spec, and the I-D says:
```
     +=====+==============+===========+==============================+
     | Bit | Name         | Reference | Assignment Notes             |
     +=====+==============+===========+==============================+
     | 7   | AE (Accurate | RFC XXXX  | Previously used as NS (Nonce |
     |     | ECN)         |           | Sum) by [RFC3540], which is  |
     |     |              |           | now historic [RFC8311]       |
     +-----+--------------+-----------+------------------------------+
```

Reply:
> Do you mean in the `[SF.PUE]` output syntax?

Yes, 'N' like 'Accurate EC**N**'. The 'W' for C**W**R and the '.' for ACK show it is not always the first character of the flag name.

Reply:
It does not work in every case, but it is nice when it works. For the avoidance of doubt, if anyone is going to suggest "Æ" for the single-character syntax, that would not work well for all users.

Reply:
Yes, keep ASCII output only.

Reply:
> To avoid confusion 'A' with ACK, could the new flag for 'Accurate ECN' be 'N'?

Other testing/tooling code (wireshark, packetdrill) is already using the "A" character here (and 0..7 when the use of the ACE field was detected in the 3WHS). Since the use of "." for ACK, as spearheaded by tcpdump, is a well recognized standard, people misinterpreting "A" here will certainly learn this fast - and remember it better for that mistake.

And I thought about Unicode, but since tcpdump is supposed to work on a high variety of platforms and terminals, introducing i8n/mapping just here is not ideal.

Reply:
Where Wireshark use "A"?  tshark gives:

```
...1 .... .... = Accurate ECN: Set
```

Reply:
> Do you mean in the `[SF.PUE]` output syntax? It is not difficult to imagine how "A" could easily confuse less experienced users. Bit 7 used to mean NS in the past, and N is the only character that appears in both "Accurate ECN" and "NS" (not in "AE" though). Or perhaps it could be just "7", which would also simplify representation of the remaining three bits. Perhaps it would be best to bring this up on the mailing list.
> 
> But the man page should be consistent with the spec, and the I-D says:
> 
> ```
>      +=====+==============+===========+==============================+
>      | Bit | Name         | Reference | Assignment Notes             |
>      +=====+==============+===========+==============================+
>      | 7   | AE (Accurate | RFC XXXX  | Previously used as NS (Nonce |
>      |     | ECN)         |           | Sum) by [RFC3540], which is  |
>      |     |              |           | now historic [RFC8311]       |
>      +-----+--------------+-----------+------------------------------+
> ```

As just mentioned, using numerals here would conflict with existing tooling/testing code, where 0..7 represents the different codepoints that the ACE field can have after the AccECN handshake. Also, this bit is canonically the 9th bit in the 12-bit wide field reserved for TCP header flags. This discussion would have been good to have last year, when the set of patches got upstreamed across a number of tools.

In any case, I updated the man page to show the full RFC abbreviations, and adjusted the indentation of the line describing the bits and their single character representation.

Reply:
> Where Wireshark use "A"? tshark gives:
> 
> ```
> ...1 .... .... = Accurate ECN: Set
> ```

See patch https://github.com/wireshark/wireshark/commit/ecefcf880121961c03d2ec05ad378c82e951e0a2

The ASCII characters used in the abbreviated decoding are "RRRACEUAPRSF" - so both ACK and AE use "A"; wireshark will however decode the full ACE field typically into one of the numericals 0..7 during the session, or use the "long" abbreviation:

    4   2.055806 192.168.20.233 → 192.168.20.216 TCP  60 0 00:0c:29:f7:80:12 08:00:27:26:77:7f  29806 → 22 [SYN, ECE, CWR, AE] Seq=0 Win=65535 Len=0 MSS=1460 WS=64 SACK_PERM TSval=155081733 TSecr=0

Reply:
Having 2 "A" when printing someting like "R--AC-UA--S-" where each "A" is at a specific place is different.
It is not the way tcpdump works.

Reply:
Yes. tcpdump (and packetdrill) have been using "." for ACK and packetdrill is using "A" for AE, and additionally 0..7 for the full ACE field (comprised of AE, CWR, ECE).

Reply:
Thank you for preparing the update, the formatting looks consistent. In order to avoid another round of discussion later, let me suggest to bring the naming/labelling aspect on the mailing list before merging these changes.

Reply:
> Use of . to indicate ACK is a peculiarity of the protocol analyzer software; the motivation behind that may or may not be lost in time,

If we look at f555c163f90c9de17ebcef8313f86404c5174ca9, which was the commit that created the tcpdump repository (from back when it used CVS, before being converted to Git), the print-tcp.c in that commit (which is the same as the tcpdump 3.4 print-tcp.c) has:

```
	if ((flags = tp->th_flags) & (TH_SYN|TH_FIN|TH_RST|TH_PUSH)) {
		if (flags & TH_SYN)
			putchar('S');
		if (flags & TH_FIN)
			putchar('F');
		if (flags & TH_RST)
			putchar('R');
		if (flags & TH_PUSH)
			putchar('P');
	} else
		putchar('.');
```

as the code that dissects the flags field.  ".", at that point, meant "this is a boring data transfer or ACK-only packet", as opposed to the initial SYN, the SYN+ACK response, an RST, *or* a "here's a PSH" packet.  (URG packets are indicated by "urg <urgent pointer value>" in the output.)

Presumably "." was chosen because it doesn't stand out much.

The commit that switched to using a table of flag values/characters and the "flag bits to string" routine was commit 9427d40c14077527b18af1bedeae39457b2981fe:

```
Author: Hannes Gredler <hannes@gredler.at>
Date:   Mon Jan 29 09:59:42 2007 +0000

    change the TCP printer to print new-style order:
      - seperate fields by comma
      - use [] for flags and options sets
      - print a trailing length field

    make use of tok2str() and bitttok2str()
    move port definitions into tcp.h
```

Presumably "." was chosen so that boring data transfers and ACK-only packets will still show up as ".".

One could argue that explicitly showing the ACK flag means that, for example, a boring data transfer that *doesn't* acknowledge anything - which, as I remember, should not happen - can be distinguished from one that does.

Reply:
I would argue the use of „.“ for the ACK flag is well established in the trade, so that a sporadic „A“ would not indroduce much ambiguity - probably more curiosity to learn what this AccECN is all about, and the larger L4S project with TCP Prague and DualQ AQM ;)